### PR TITLE
Implement Pluggable Discovery protocol v1

### PR DESCRIPTION
**Pull request by cmaglie**
Ongoing discussion here: https://github.com/arduino/tooling-rfcs/pull/2

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,7 +8,35 @@ Install a recent go enviroment (>=13.0) and run `go build`. The executable `seri
...
+
+for example:
+
+`HELLO 1 "Arduino IDE"`
+
+or:
+
+`HELLO 1 "arduino-cli"`
+
+in this case the protocol version requested by the clien is `1` (at the moment of writing there were no other revisions of the protocol).
```
Comment (per1234) [inline README.md:28]:
```suggestion
in this case the protocol version requested by the client is `1` (at the moment of writing there were no other revisions of the protocol).
```

**Review by per1234 — CHANGES_REQUESTED** [review]

File: main.go [hunk 2]
```diff
@@ -94,68 +149,67 @@ type listOutputJSON struct {
...
+		portJSON.Properties.Set("pid", "0x"+port.PID)
+		portJSON.Properties.Set("serialNumber", port.SerialNumber)
 	}
 	return portJSON
 }
 
-type messageOutputJSON struct {
+type helloMessageJSON struct {
+	EventType       string `json:"eventType"`
+	ProtocolVersion string `json:"protocolVersion"`
```
Comment (silvanocerza) [inline main.go:190]: Why not an `int`?

Comment (cmaglie) [reply main.go:190]: mmmhhhhh... I need to change the specification, but you're right, better to be `int`

File: main.go [hunk 3]
```diff
@@ -42,48 +44,101 @@ func main() {
...
+				Message:   err.Error(),
+			})
 			os.Exit(1)
 		}
-		cmd = strings.ToUpper(strings.TrimSpace(cmd))
+		split := strings.Split(fullCmd, " ")
+		cmd := strings.ToUpper(strings.TrimSpace(split[0]))
 		switch cmd {
+		case "HELLO":
+			re := regexp.MustCompile(`(\d+) "([^"]+)"`)
```
Comment (silvanocerza) [inline main.go:60]: I think we should avoid compiling this `regexp` for every `HELLO` message by compiling it outside the loop.

File: main.go [hunk 4]
```diff
@@ -42,48 +44,101 @@ func main() {
...
+			if len(matches) != 3 {
+				output(&genericMessageJSON{
+					EventType: "command_error",
+					Error:     true,
+					Message:   "Invalid HELLO command",
+				})
+				continue
+			}
+			// userAgent := matches[2]
+			// reqProtocolVersion, err := strconv.ParseUint(matches[1], 10, 64)
```
Comment (silvanocerza) [inline main.go:71]: Remove this if they're not necessary.

File: main.go [hunk 5]
```diff
@@ -42,48 +44,101 @@ func main() {
...
+			_, err := strconv.ParseUint(matches[1], 10, 64)
+			if err != nil {
+				output(&genericMessageJSON{
+					EventType: "command_error",
+					Error:     true,
+					Message:   "Invalid protocol version: " + matches[2],
+				})
+			}
+			// fmt.Println("User agent:", userAgent)
+			// fmt.Println("Req. Protocol version:", reqProtocolVersion)
```
Comment (silvanocerza) [inline main.go:81]: These should be removed too or logged if necessary.

File: main.go (hunk 3 above)
Comment (cmaglie) [reply main.go:60]: Moving `regexp.Compile` outside the loop will keep the compiled regexp in RAM forever while keeping it inside the loop will consume CPU cycles to compile the regexp every time the `HELLO` command is executed. Since the `HELLO` command is supposed to be called just once at the beginning, the `regexp.MustCompile` is therefore called just once: moving it outside the loop doesn't give us any advantage (the regexp will be compiled one time in both cases).

BTW the amount of RAM or CPU we are saving in a case or another is negligible, so I will give priority to readability: moving the regexp outside the loop will move it away from his main context (the `HELLO` command handler), reducing code readability.

In conclusion: I will not move it outside the loop :-)

Comment (silvanocerza) [reply main.go:60]: Makes sense. :+1:

**Review by silvanocerza — APPROVED** [review]
